controller/volume_controller.go

```diff
@@ -1242,63 +1220,22 @@ func isDataLocalityDisabled(v *longhorn.Volume) bool {
 	return string(v.Spec.DataLocality) == "" || v.Spec.DataLocality == types.DataLocalityDisabled
 }
 
-// hasLocalReplicaOnSameNodeAsEngine returns whether there exist a replica on the same node as engine
+// hasLocalReplicaOnSameNodeAsEngine returns true if one of the following condition is satisfied:
+// 1. there exist a replica on the same node as engine
+// 2. there exist a replica with HardNodeAffinity set to engine's NodeID
 func hasLocalReplicaOnSameNodeAsEngine(e *longhorn.Engine, rs map[string]*longhorn.Replica) bool {
 	for _, r := range rs {
-		if r.Spec.NodeID == e.Spec.NodeID {
+		if e.Spec.NodeID != "" && r.Spec.NodeID == e.Spec.NodeID || r.Spec.HardNodeAffinity == e.Spec.NodeID {
```

Comment:
Seems the precedence is wrong here? Should it be `e.Spec.NodeID != "" && (r.Spec.NodeID == e.Spec.NodeID || r.Spec.HardNodeAffinity == e.Spec.NodeID)` ?

Reply:
Oops. My bad. I am sorry!